From 41b9be6e64ce92553d84a1c255d049fc42636fef Mon Sep 17 00:00:00 2001
From: Jean-Yves Avenard <jyavenard@mythtv.org>
Date: Sat, 19 Jul 2014 03:39:27 +1000
Subject: [PATCH 3/3] =?UTF-8?q?frontend:=20add=20search=20abilities=20to=20=E2?=
 =?UTF-8?q?=80=9Cchange=20recording=20metadata=E2=80=9D=20screen.?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This will allow to search for inetref and set season/episode as well as subtitle and description.
Themes will have to be updated to enjoy this functionality by adding a “query_button” button.
---
 mythtv/programs/mythfrontend/playbackbox.cpp |  158 +++++++++++++++++++++++++-
 mythtv/programs/mythfrontend/playbackbox.h   |   13 ++-
 2 files changed, 166 insertions(+), 5 deletions(-)

diff --git a/mythtv/programs/mythfrontend/playbackbox.cpp b/mythtv/programs/mythfrontend/playbackbox.cpp
index a1e1e91..a812968 100644
--- a/mythtv/programs/mythfrontend/playbackbox.cpp
+++ b/mythtv/programs/mythfrontend/playbackbox.cpp
@@ -20,6 +20,8 @@
 #include "mythuitextedit.h"
 #include "mythuispinbox.h"
 #include "mythdialogbox.h"
+#include "mythprogressdialog.h"
+#include "mythuimetadataresults.h"
 #include "recordinginfo.h"
 #include "recordingrule.h"
 #include "mythuihelper.h"
@@ -5257,11 +5259,14 @@ void PasswordChange::SendResult()
 ////////////////////////////////////////////////////////
 
 RecMetadataEdit::RecMetadataEdit(MythScreenStack *lparent, ProgramInfo *pginfo)
-                : MythScreenType(lparent, "recmetadataedit"),
-                    m_progInfo(pginfo)
+  : MythScreenType(lparent, "recmetadataedit"),
+    m_titleEdit(NULL),      m_subtitleEdit(NULL),   m_descriptionEdit(NULL),
+    m_inetrefEdit(NULL),    m_seasonSpin(NULL),     m_episodeSpin(NULL),
+    m_busyPopup(NULL),      m_queryButton(NULL),
+    m_progInfo(pginfo)
 {
-    m_titleEdit = m_subtitleEdit = m_descriptionEdit = m_inetrefEdit = NULL;
-    m_seasonSpin = m_episodeSpin = NULL;
+    m_popupStack = GetMythMainWindow()->GetStack("popup stack");
+    m_metadataFactory = new MetadataFactory(this);
 }
 
 bool RecMetadataEdit::Create()
@@ -5276,6 +5281,7 @@ bool RecMetadataEdit::Create()
     m_seasonSpin = dynamic_cast<MythUISpinBox*>(GetChild("season"));
     m_episodeSpin = dynamic_cast<MythUISpinBox*>(GetChild("episode"));
     MythUIButton *okButton = dynamic_cast<MythUIButton*>(GetChild("ok"));
+    m_queryButton = dynamic_cast<MythUIButton*>(GetChild("query_button"));
 
     if (!m_titleEdit || !m_subtitleEdit || !m_inetrefEdit || !m_seasonSpin ||
         !m_episodeSpin || !okButton)
@@ -5302,6 +5308,10 @@ bool RecMetadataEdit::Create()
     m_episodeSpin->SetValue(m_progInfo->GetEpisode());
 
     connect(okButton, SIGNAL(Clicked()), SLOT(SaveChanges()));
+    if (m_queryButton)
+    {
+        connect(m_queryButton, SIGNAL(Clicked()), SLOT(PerformQuery()));
+    }
 
     BuildFocusList();
 
@@ -5329,6 +5339,146 @@ void RecMetadataEdit::SaveChanges()
     Close();
 }
 
+void RecMetadataEdit::PerformQuery()
+{
+    if (m_busyPopup)
+        return;
+
+    m_busyPopup = new MythUIBusyDialog(tr("Trying to manually find this "
+                                          "recording online..."),
+                                       m_popupStack,
+                                       "metaoptsdialog");
+
+    if (m_busyPopup->Create())
+        m_popupStack->AddScreen(m_busyPopup);
+
+    MetadataLookup *lookup = new MetadataLookup();
+    lookup->SetStep(kLookupSearch);
+    lookup->SetType(kMetadataRecording);
+    LookupType type = GuessLookupType(m_inetrefEdit->GetText());
+
+    if (type == kUnknownVideo)
+    {
+        if (m_seasonSpin->GetIntValue() == 0 &&
+            m_episodeSpin->GetIntValue() == 0 &&
+            m_subtitleEdit->GetText().isEmpty())
+        {
+            lookup->SetSubtype(kProbableMovie);
+        }
+        else
+        {
+            lookup->SetSubtype(kProbableTelevision);
+        }
+    }
+    else
+    {
+        // we could determine the type from the inetref
+        lookup->SetSubtype(type);
+    }
+    lookup->SetAllowGeneric(true);
+    lookup->SetHandleImages(false);
+    lookup->SetHost(gCoreContext->GetMasterHostName());
+    lookup->SetTitle(m_titleEdit->GetText());
+    lookup->SetSubtitle(m_subtitleEdit->GetText());
+    lookup->SetInetref(m_inetrefEdit->GetText());
+    lookup->SetCollectionref(m_inetrefEdit->GetText());
+    lookup->SetSeason(m_seasonSpin->GetIntValue());
+    lookup->SetEpisode(m_episodeSpin->GetIntValue());
+    lookup->SetAutomatic(false);
+
+    m_metadataFactory->Lookup(lookup);
+}
+
+void RecMetadataEdit::QueryComplete(MetadataLookup *lookup)
+{
+    if (!lookup)
+        return;
+
+    m_inetrefEdit->SetText(lookup->GetInetref());
+    m_seasonSpin->SetValue(lookup->GetSeason());
+    m_episodeSpin->SetValue(lookup->GetEpisode());
+    if (!lookup->GetSubtitle().isEmpty())
+    {
+        m_descriptionEdit->SetText(lookup->GetSubtitle());
+    }
+    if (!lookup->GetDescription().isEmpty())
+    {
+        m_descriptionEdit->SetText(lookup->GetDescription());
+    }
+}
+
+void RecMetadataEdit::OnSearchListSelection(RefCountHandler<MetadataLookup> lookup)
+{
+    QueryComplete(lookup);
+}
+
+void RecMetadataEdit::customEvent(QEvent *levent)
+{
+    if (levent->type() == MetadataFactoryMultiResult::kEventType)
+    {
+        if (m_busyPopup)
+        {
+            m_busyPopup->Close();
+            m_busyPopup = NULL;
+        }
+
+        MetadataFactoryMultiResult *mfmr = dynamic_cast<MetadataFactoryMultiResult*>(levent);
+
+        if (!mfmr)
+            return;
+
+        MetadataLookupList list = mfmr->results;
+
+        MetadataResultsDialog *resultsdialog =
+            new MetadataResultsDialog(m_popupStack, list);
+
+        connect(resultsdialog, SIGNAL(haveResult(RefCountHandler<MetadataLookup>)),
+                SLOT(OnSearchListSelection(RefCountHandler<MetadataLookup>)),
+                Qt::QueuedConnection);
+
+        if (resultsdialog->Create())
+            m_popupStack->AddScreen(resultsdialog);
+    }
+    else if (levent->type() == MetadataFactorySingleResult::kEventType)
+    {
+        if (m_busyPopup)
+        {
+            m_busyPopup->Close();
+            m_busyPopup = NULL;
+        }
+
+        MetadataFactorySingleResult *mfsr = dynamic_cast<MetadataFactorySingleResult*>(levent);
+
+        if (!mfsr && !mfsr->result)
+            return;
+
+        QueryComplete(mfsr->result);
+    }
+    else if (levent->type() == MetadataFactoryNoResult::kEventType)
+    {
+        if (m_busyPopup)
+        {
+            m_busyPopup->Close();
+            m_busyPopup = NULL;
+        }
+
+        MetadataFactoryNoResult *mfnr = dynamic_cast<MetadataFactoryNoResult*>(levent);
+
+        if (!mfnr)
+            return;
+
+        QString title = tr("No match found for this recording. You can "
+                           "try entering a TVDB/TMDB number, season, and "
+                           "episode manually.");
+
+        MythConfirmationDialog *okPopup =
+            new MythConfirmationDialog(m_popupStack, title, false);
+
+        if (okPopup->Create())
+            m_popupStack->AddScreen(okPopup);
+    }
+}
+
 //////////////////////////////////////////
 
 HelpPopup::HelpPopup(MythScreenStack *lparent)
diff --git a/mythtv/programs/mythfrontend/playbackbox.h b/mythtv/programs/mythfrontend/playbackbox.h
index e86e6bf..d4bafa2 100644
--- a/mythtv/programs/mythfrontend/playbackbox.h
+++ b/mythtv/programs/mythfrontend/playbackbox.h
@@ -22,6 +22,7 @@ using namespace std;
 #include "tv_play.h"
 
 #include "mythscreentype.h"
+#include "metadatafactory.h"
 
 // mythfrontend
 #include "schedulecommon.h"
@@ -44,6 +45,7 @@ class MythUITextEdit;
 class MythUIButton;
 class MythDialogBox;
 class MythMenu;
+class MythUIBusyDialog;
 
 typedef QMap<QString,ProgramList>       ProgramMap;
 typedef QMap<QString,QString>           Str2StrMap;
@@ -533,16 +535,25 @@ class RecMetadataEdit : public MythScreenType
 
   protected slots:
     void SaveChanges(void);
+    void PerformQuery(void);
+    void OnSearchListSelection(RefCountHandler<MetadataLookup> lookup);
 
   private:
+    void customEvent(QEvent *event);
+    void QueryComplete(MetadataLookup *lookup);
+
     MythUITextEdit     *m_titleEdit;
     MythUITextEdit     *m_subtitleEdit;
     MythUITextEdit     *m_descriptionEdit;
     MythUITextEdit     *m_inetrefEdit;
     MythUISpinBox      *m_seasonSpin;
     MythUISpinBox      *m_episodeSpin;
+    MythUIBusyDialog   *m_busyPopup;
+    MythUIButton       *m_queryButton;
 
-    ProgramInfo *m_progInfo;
+    ProgramInfo        *m_progInfo;
+    MythScreenStack    *m_popupStack;
+    MetadataFactory    *m_metadataFactory;
 };
 
 class HelpPopup : public MythScreenType
-- 
1.7.10.2

